# Real Estate Market Analysis with Python Project

### Data Preprocessing
- The initial stage involves cleaning and preparing the data for real estate analysis, including handling missing values, correcting inconsistencies, and transforming data types if necessary. You must clean and preprocess the customers and property tables, ensure column names are in order and that missing values appear correctly, and apply any column data type changes you see fit.

- Finally, you should combine the properties  and customers tables into one unified real estate dataset using the shared customer_id column. You must consolidate inconsistencies or missing values to perform the following analysis and obtain the correct results.

### Properties and Customers

You must review the list, preprocess, and clean both datasets accordingly.

1 - Descriptive statistics: You can start the analysis with descriptive statistics of the data and check for missing values.

2 - Datatypes: Evaluate the datatypes of the columns and decide whether some of the datatypes need to be changed.

3 - Column names: Check if there is an issue with any column names and rename them if necessary.

4 - Categorical to numerical: Change categorical values to numeric when possible and needed. Use the 0 1 convention when mapping.

5 - Case: If there are inconsistencies with Capitol and lowercases, unify them using the lowercase convention.

6 - Missing values: Ensure missing values are correctly indicated.

7 - Date variables: Make sure you handle and transform the date variables as a date.

### Combining the Two Datasets

- This phase aims to merge our two cleaned datasets (properties  and customers) into a comprehensive dataset.

1 - Preliminary Checks. Visually inspect the two datasets and decide which variable to merge. As both datasets share only column – you must opt for the customer_id  column.

2 - Initial Merge Attempt. You can leverage the pandas' functionality to combine the two tables without initial preprocessing. Think about what kind of join would make sense in the context of the given problem.

3 - Resolving Merge Issues. If the initial attempt to merge the data does not yield the correct data frame, you must examine and preprocess the data further. Think about which variable is likely causing the merge issue and review it closely- it makes sence to check the customer_id column. You must examine the values of the ids for both properties  and customers datasets. You will notice that there are unneccessary spaces in one of the datasets which you must remove. After that you should be able to merge successfully.

##### Sanity check: You should obtain a pandas data frame of 267 rows and 19 columns.

##### Imports

In [161]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
from sklearn.preprocessing import LabelEncoder

In [162]:
customers = pd.read_csv('customers.csv')
properties = pd.read_csv('properties.csv')

In [163]:
customers.head()

Unnamed: 0 customerid      entity     name  surname  birth_date  sex  \
0           0       C0110  Individual   Kareem      Liu   5/11/1968    F   
1           1       C0010  Individual  Trystan  Oconnor  11/26/1962    M   
2           2       C0132  Individual     Kale      Gay    4/7/1959    M   
3           3       C0137  Individual  Russell    Gross  11/25/1959    M   
4           4       C0174     Company  Marleez       Co         NaN  NaN   

  country       state     purpose  deal_satisfaction mortgage   source  
0     USA  California        Home                  4      Yes  Website  
1     USA  California        Home                  1       No  Website  
2     USA  California        Home                  4      Yes   Agency  
3     USA  California        Home                  5       No  Website  
4    USA   California  Investment                  5       No  Website

In [164]:
properties.head()

Unnamed: 0   id  building  date_sale       type  property#     area  \
0           0  1030         1  11/1/2005  Apartment         30   743.09   
1           1  1029         1  10/1/2005  Apartment         29   756.21   
2           2  2002         2   7/1/2007  Apartment          2   587.28   
3           3  2031         2  12/1/2007  Apartment         31  1604.75   
4           4  1049         1  11/1/2004  Apartment         49  1375.45   

          price  status customerid  
0  $246,172.68    Sold      C0028   
1  $246,331.90    Sold      C0027   
2  $209,280.91    Sold      C0112   
3  $452,667.01    Sold      C0160   
4  $467,083.31    Sold      C0014

In [165]:
print(customers.columns)


Index(['Unnamed: 0', 'customerid', 'entity', 'name', 'surname', 'birth_date',
       'sex', 'country', 'state', 'purpose', 'deal_satisfaction', 'mortgage',
       'source'],
      dtype='object')


In [166]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162 entries, 0 to 161
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Unnamed: 0         162 non-null    int64 
 1   customerid        162 non-null    object
 2   entity             162 non-null    object
 3   name               162 non-null    object
 4   surname            162 non-null    object
 5   birth_date         155 non-null    object
 6   sex                155 non-null    object
 7   country            162 non-null    object
 8   state              154 non-null    object
 9   purpose            162 non-null    object
 10  deal_satisfaction  162 non-null    int64 
 11  mortgage           162 non-null    object
 12  source             162 non-null    object
dtypes: int64(2), object(11)
memory usage: 16.6+ KB


In [167]:
customers.describe()

Unnamed: 0  deal_satisfaction
count  162.000000         162.000000
mean    80.500000           3.456790
std     46.909487           1.333276
min      0.000000           1.000000
25%     40.250000           3.000000
50%     80.500000           4.000000
75%    120.750000           5.000000
max    161.000000           5.000000

In [168]:
customers.shape

(162, 13)

### Data Preparation

In [169]:
# Drop the first column 'Unnamed: 0' from both the dataframes
customers.drop(columns=['Unnamed: 0','name','surname'], axis=1, inplace=True)

### Checking Datatypes

In [170]:
print(customers.columns)

Index(['customerid', 'entity', 'birth_date', 'sex', 'country', 'state',
       'purpose', 'deal_satisfaction', 'mortgage', 'source'],
      dtype='object')


In [171]:
customers.dtypes

customerid          object
entity               object
birth_date           object
sex                  object
country              object
state                object
purpose              object
deal_satisfaction     int64
mortgage             object
source               object
dtype: object

In [172]:
# Convert the data types of the columns to their appropriate types 'strings'
customers['customerid'] = customers['customerid'].astype('string')
customers['entity'] = customers['entity'].astype('string')
customers['sex'] = customers['sex'].astype('string')
customers['country'] = customers['country'].astype('string')
customers['state'] = customers['state'].astype('string')
customers['purpose'] = customers['purpose'].astype('string')
customers['mortgage'] = customers['mortgage'].astype('string')
customers['source'] = customers['source'].astype('string')

In [173]:
# Convert the data types of the columns to their appropriate types 'date'
customers['birth_date'] = customers['birth_date'].astype('datetime64[ns]')

In [174]:
customers.dtypes

customerid                  string
entity                       string
birth_date           datetime64[ns]
sex                          string
country                      string
state                        string
purpose                      string
deal_satisfaction             int64
mortgage                     string
source                       string
dtype: object

### Column names

In [175]:
# rename columns
customers.rename(columns={'customerid': 'customerid'}, inplace=True)
customers.rename(columns={'birth_date':'birthdate'}, inplace=True)

### Categorical to numerical

### Case: If there are inconsistencies with Capitol and lowercases, unify them using the lowercase convention.


using the lowercase method

In [176]:
categorical_cols = ['entity',  'sex', 'country', 'state',
       'purpose', 'mortgage', 'source']

In [177]:
for col in categorical_cols:
  customers[col] = customers[col].str.lower()

In [178]:
# using label encoder to encode categorical columns
label_encoder = LabelEncoder()

##### we have first to check the unoque values of each column

##### 1 - Entity column

In [179]:
print(customers['entity'].unique())

<StringArray>
['individual', 'company']
Length: 2, dtype: string


In [197]:
customers['entity'] = label_encoder.fit_transform(customers['entity'])
print(customers['entity'].unique())

[1 0]


2 - Sex column

In [180]:
print(customers['sex'].unique())

<StringArray>
['f', 'm', <NA>]
Length: 3, dtype: string


##### Imputation of NA values

In [181]:
customers['sex'].fillna('Unknown', inplace=True)

In [182]:
customers['sex'] = label_encoder.fit_transform(customers['sex'])
print(customers['sex'].unique())

[1 2 0]


3 - country column

1.   Remove leading and trailing whitespace from all values in the 'country' column.
2.   Convert all values to lowercase to ensure uniform capitalization.


In [184]:
customers['country'] = customers['country'].str.strip().str.lower()

In [185]:
print(customers['country'].unique())

<StringArray>
['usa', 'canada', 'germany', 'belgium', 'mexico', 'russia', 'uk', 'denmark']
Length: 8, dtype: string


In [186]:
customers['country'] = label_encoder.fit_transform(customers['country'])
print(customers['country'].unique())

[7 1 3 0 4 5 6 2]


4 - state column

In [187]:
print(customers['state'].unique())

<StringArray>
['california',         <NA>,     'oregon',    'arizona',     'nevada',
   'colorado',       'utah',   'virginia',    'wyoming',     'kansas']
Length: 10, dtype: string


##### Imputation of NA values

In [188]:
customers['state'].fillna('unknown',inplace=True)

In [189]:
customers['state'] = label_encoder.fit_transform(customers['state'])
print(customers['state'].unique())

[1 6 5 0 4 2 7 8 9 3]


5 - purpose column

In [190]:
print(customers['purpose'].unique())

<StringArray>
['home', 'investment']
Length: 2, dtype: string


In [191]:
customers['purpose'] = label_encoder.fit_transform(customers['purpose'])
print(customers['purpose'].unique())

[0 1]


6 - mortgage column

In [192]:
print(customers['mortgage'].unique())

<StringArray>
['yes', 'no']
Length: 2, dtype: string


In [193]:
customers['mortgage'] = label_encoder.fit_transform(customers['mortgage'])
print(customers['mortgage'].unique())

[1 0]


7 - source column

In [194]:
print(customers['source'].unique())

<StringArray>
['website', 'agency', 'client']
Length: 3, dtype: string


In [195]:
customers['source'] = label_encoder.fit_transform(customers['source'])
print(customers['source'].unique())

[2 0 1]


In [198]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162 entries, 0 to 161
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   customerid         162 non-null    string        
 1   entity             162 non-null    int64         
 2   birthdate          155 non-null    datetime64[ns]
 3   sex                162 non-null    int64         
 4   country            162 non-null    int64         
 5   state              162 non-null    int64         
 6   purpose            162 non-null    int64         
 7   deal_satisfaction  162 non-null    int64         
 8   mortgage           162 non-null    int64         
 9   source             162 non-null    int64         
dtypes: datetime64[ns](1), int64(8), string(1)
memory usage: 12.8 KB


In [199]:
customers.head()

customerid  entity  birthdate  sex  country  state  purpose  \
0      C0110       1 1968-05-11    1        7      1        0   
1      C0010       1 1962-11-26    2        7      1        0   
2      C0132       1 1959-04-07    2        7      1        0   
3      C0137       1 1959-11-25    2        7      1        0   
4      C0174       0        NaT    0        7      1        1   

   deal_satisfaction  mortgage  source  
0                  4         1       2  
1                  1         0       2  
2                  4         1       0  
3                  5         0       2  
4                  5         0       2

In [200]:
# For customers dataframe
customers.isnull().sum()

customerid           0
entity               0
birthdate            7
sex                  0
country              0
state                0
purpose              0
deal_satisfaction    0
mortgage             0
source               0
dtype: int64

In [201]:
missing_percentage = (customers.isnull().mean() * 100).round(2)
print(missing_percentage)

customerid           0.00
entity               0.00
birthdate            4.32
sex                  0.00
country              0.00
state                0.00
purpose              0.00
deal_satisfaction    0.00
mortgage             0.00
source               0.00
dtype: float64


##### Droping the rows that contains missing values because it is with 4.32 % of the dataset.

In [202]:
customers.dropna(subset=['birthdate'], inplace=True)

In [203]:
customers.shape

(155, 10)

In [88]:
# For properties dataframe
properties.isnull().sum()

Unnamed: 0     0
id            0
building       0
date_sale      0
type           0
property#      0
area           0
price          0
status         0
customerid    72
dtype: int64